cirq-core/cirq/circuits/frozen_circuit.py

```diff
@@ -55,9 +55,20 @@ def __init__(
     def moments(self) -> Sequence['cirq.Moment']:
         return self._moments
 
-    def __hash__(self):
+    @_compat.cached_method
+    def __hash__(self) -> int:
+        # Explicitly cached for performance
         return hash((self.moments,))
 
+    def __getstate__(self):
+        # Don't save hash when pickling; see #3777.
+        state = self.__dict__
+        hash_cache = _compat._method_cache_name(self.__hash__)
+        if hash_cache in state:
+            state = state.copy()
+            del state[hash_cache]
```

Comment:
Optional suggestion:

```python
state = self.__dict__.copy()
hash_cache = _compat._method_cache_name(self.__hash__)
state.pop(hash_cache, None)
return state
```

Rational: This is consistent in always returning a copy of `self.__dict__`. Also, it avoids branching.

(If you choose to do this, apply to GridQubit, too.)

Reply:
AFAICT from the docs on [`__getstate__`](https://docs.python.org/3/library/pickle.html#object.__getstate__) it's not necessary to return a copy, since the pickle infrastructure does not modify the returned object. In fact, pickle will use `self.__dict__` directly if `__getstate__` is not defined.